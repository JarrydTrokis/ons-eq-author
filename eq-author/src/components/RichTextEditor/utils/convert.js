import { convertToHTML, convertFromHTML } from "draft-convert";
import { EditorState } from "draft-js";

export const toHTML = entityMap => {
  const entityToHTML = (entity, originalText) => {
    const mapper = entityMap[entity.type];
    return mapper ? mapper(entity) : originalText;
  };

  const convert = convertToHTML({ entityToHTML });

  return editorState => convert(editorState.getCurrentContent());
};

export const fromHTML = nodeToFn => {
  const htmlToEntity = (nodeName, ...otherArgs) => {
    const entity = Object.entries(nodeToFn)
      .filter(([name]) => name === nodeName)
      .map(([, fn]) => fn(nodeName, ...otherArgs))
      .find(result => result);

    return entity || null;
  };

  const convert = convertFromHTML({ htmlToEntity });

  return (html, decorator) =>
    EditorState.createWithContent(convert(html), decorator);
};
