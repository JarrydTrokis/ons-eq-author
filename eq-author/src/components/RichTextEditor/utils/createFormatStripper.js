import { mapKeys } from "lodash";
import { filterEditorState } from "draftjs-filters";
import { filterConfig as pipedFilterConfig } from "components/RichTextEditor/entities/PipedValue";
import { filterConfig as linkFilterConfig } from "components/RichTextEditor/LinkPlugin";

const mapper = {
  bold: { format: "BOLD", type: "styles" },
  emphasis: { format: "ITALIC", type: "styles" },
  list: { format: "unordered-list-item", type: "blocks" },
  heading: { format: "header-two", type: "blocks" },
};

// Custom filter function to replace multiple spaces with a single space
function replaceMultipleSpaces(contentState) {
  const text = contentState.getText();
  const transformedText = text.replace(/\s+/g, " "); // Replace multiple spaces with a single space
  return contentState.merge({
    text: transformedText,
  });
}

export default function createFormatStripper(controls) {
  const filterConfiguration = {
    blocks: [],
    styles: [],
    entities: [pipedFilterConfig, linkFilterConfig],
    maxNesting: 0,
    whitespacedCharacters: [replaceMultipleSpaces],
  };

  mapKeys(controls, (value, key) => {
    if (mapper[key] && controls[key]) {
      filterConfiguration[mapper[key].type].push(mapper[key].format);
    }
  });

  return (editorState) => filterEditorState(filterConfiguration, editorState);
}
