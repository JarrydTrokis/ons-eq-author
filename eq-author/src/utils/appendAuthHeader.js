export default headers => {
  const accessToken = localStorage.getItem("accessToken");
  const returnedHeaders = {
    ...headers
  };

  if (accessToken) {
    returnedHeaders.authorization = `Bearer ${accessToken}`;
  }

  return {
    headers: {
      ...returnedHeaders
    }
  };
};
