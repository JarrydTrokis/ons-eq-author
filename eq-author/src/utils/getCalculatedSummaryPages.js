const insertSection = (sections, section, folder, page, answers) => {
  sections.push({
    ...section,
    folders: [],
  });
  sections = insertFolder(sections, folder, page, answers);
  return sections;
};

const insertFolder = (sections, folder, page, answers) => {
  sections[sections.length - 1].folders.push({
    ...folder,
    pages: [],
  });
  sections = insertPage(sections, page, answers);
  return sections;
};

const insertPage = (sections, page, answers) => {
  sections[sections.length - 1].folders[
    sections[sections.length - 1].folders.length - 1
  ].pages.push({
    ...page,
    answers,
  });
  return sections;
};

const getCalculatedSummaryPages = (questionnaire, sectionId) => {
  const sections = [];
  if (questionnaire?.sections) {
    for (const section of questionnaire?.sections) {
      if (section.id === sectionId) {
        return sections;
      }
      for (const folder of section?.folders) {
        for (const page of folder?.pages) {
          const answers = page?.answers;
          if (page?.pageType === "CalculatedSummaryPage") {
            if (sections[sections.length - 1]?.id === section?.id) {
              insertPage(sections, page, answers);
            } else {
              insertSection(sections, section, folder, page, answers);
            }
          }
        }
      }
    }
  }
  return sections;
};

export default getCalculatedSummaryPages;
