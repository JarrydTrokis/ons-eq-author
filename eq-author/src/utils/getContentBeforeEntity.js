import { remove } from "lodash";

import isListCollectorPageType from "utils/isListCollectorPageType";

import { MUTUALLY_EXCLUSIVE } from "constants/answer-types";

const identity = (x) => x;

const getContentBeforeEntity = (
  questionnaire,
  id,
  preprocessAnswers,
  includeTarget,
  expressionGroup
) => {
  const sections = [];

  for (const section of questionnaire.sections) {
    if (section.id === id) {
      return sections;
    }

    sections.push({
      ...section,
      folders: [],
    });

    for (const folder of section.folders) {
      if (folder.id === id) {
        return sections;
      }

      sections[sections.length - 1].folders.push({
        ...folder,
        pages: [],
      });

      for (const page of folder.pages) {
        if (page.id === id && !includeTarget) {
          return sections;
        }

        let answers =
          !isListCollectorPageType(page.pageType) &&
          (page?.answers?.flatMap(preprocessAnswers) || []);

        if (expressionGroup?.operator === "And") {
          expressionGroup.expressions.forEach((expression) => {
            if (expression?.left?.page?.id) {
              answers = answers.filter((answer) => {
                if (expression.left.page.id === answer.page.id) {
                  if (expression.left.type === MUTUALLY_EXCLUSIVE) {
                    return false;
                  } else {
                    return answer.type !== MUTUALLY_EXCLUSIVE;
                  }
                }
                return true;
              });
            }
          });
        }

        if (answers.length) {
          sections[sections.length - 1].folders[
            sections[sections.length - 1].folders.length - 1
          ].pages.push({
            ...page,
            answers,
          });
        }

        if (
          (page.id === id && includeTarget) ||
          page?.confirmation?.id === id
        ) {
          return sections;
        }
      }
    }
  }

  return sections;
};

export const stripEmpty = (sections) => {
  sections.forEach(({ folders }) =>
    remove(folders, (folder) => !folder.pages.length)
  );
  remove(sections, (section) => !section.folders.length);
  return sections;
};

export default ({
  questionnaire,
  id,
  preprocessAnswers = identity,
  includeTargetPage = false,
  expressionGroup,
} = {}) => {
  if (!questionnaire || !id || questionnaire?.introduction?.id === id) {
    return [];
  }

  return stripEmpty(
    getContentBeforeEntity(
      questionnaire,
      id,
      preprocessAnswers,
      includeTargetPage,
      expressionGroup
    ).filter(({ folders }) => folders.length)
  );
};
