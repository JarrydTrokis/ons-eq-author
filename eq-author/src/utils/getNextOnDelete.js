import { findIndex } from "lodash";

export default (items, id) => {
  const index = findIndex(items, { id });
  let nextIndex;

  if (index === 0) {
    nextIndex = 0;
  } else {
    nextIndex = index - 1;
  }

  return nextIndex;
};
