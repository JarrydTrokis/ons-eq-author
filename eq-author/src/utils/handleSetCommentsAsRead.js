import { useEffect } from "react";
import UPDATE_COMMENTS_AS_READ from "../graphql/updateCommentsAsRead.graphql";
import { useMutation } from "@apollo/react-hooks";

// Function begins with capital letter - allows useEffect to be used here
const HandleSetCommentsAsRead = (pageId, userId) => {
  const [updateCommentsAsRead] = useMutation(UPDATE_COMMENTS_AS_READ);

  // https://stackoverflow.com/questions/66404382/how-to-detect-route-changes-using-react-router-in-react
  useEffect(() => {
    return function cleanup() {
      updateCommentsAsRead({
        variables: {
          input: {
            pageId,
            userId,
          },
        },
      });
    };
  }, [updateCommentsAsRead, pageId, userId]);
};

export default HandleSetCommentsAsRead;
