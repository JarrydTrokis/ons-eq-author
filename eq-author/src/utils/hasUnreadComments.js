const hasUnreadComments = (comments, userId) => {
  if (!comments) {
    return false;
  }
  comments.forEach((comment) => {
    if (!comment.readBy.some((id) => userId === id)) {
      return true;
    }
    comment.replies.forEach((reply) => {
      if (!reply.readBy.some((id) => userId === id)) {
        return true;
      }
    });
  });
  return false;
};

export default hasUnreadComments;
