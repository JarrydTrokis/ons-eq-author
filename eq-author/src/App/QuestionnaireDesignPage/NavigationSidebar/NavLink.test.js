import React from "react";
import { render } from "tests/utils/rtl";
import NavLink, { activeClassName } from "./NavLink";

describe("NavLink", () => {
  const props = {
    to: "/page-1",
    title: "Page 1",
    icon: () => <svg />,
    isActive: () => true,
  };

  it("should have the active class name when on current page", () => {
    const { getByTestId } = render(
      <NavLink {...props} isActive={() => true}>
        Page 1
      </NavLink>
    );
    expect(getByTestId("nav-link")).toHaveClass(activeClassName);
  });

  it("should not have the active class when not on the current page", () => {
    const { getByTestId } = render(
      <NavLink {...props} isActive={() => false}>
        Page 1
      </NavLink>
    );
    expect(getByTestId("nav-link")).not.toHaveClass(activeClassName);
  });

  it("should show a badge when it has errors", () => {
    const { getByTestId } = render(
      <NavLink {...props} errorCount={9999}>
        Page 1
      </NavLink>
    );
    expect(getByTestId("nav-link")).toHaveTextContent(9999);
  });

  it("should show a badge when section accordion is closed and has errors within the questions", () => {
    const { getByTestId } = render(
      <NavLink
        {...props}
        errorCount={9999}
        isOpen={false}
        isSection
        sectionTotalErrors={1}
      >
        Page 1
      </NavLink>
    );
    expect(getByTestId("badge-NoCount-closed")).toBeTruthy();
  });

  it("should show a small badge when section accordion is closed and section has errors", () => {
    const { getByTestId } = render(
      <NavLink
        {...props}
        errorCount={9999}
        isOpen={false}
        isSection
        sectionTotalErrors={1}
      >
        Page 1
      </NavLink>
    );
    expect(getByTestId("badge-NoCount-closed")).toBeTruthy();
  });
});
