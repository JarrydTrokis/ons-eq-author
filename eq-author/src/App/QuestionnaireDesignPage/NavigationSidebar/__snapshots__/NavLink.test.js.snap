// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`NavLink should render 1`] = `
<NavLink__Link
  activeClassName="active"
  isActive={[Function]}
  title="Page 1"
  to="/page-1"
>
  <IconText
    hideText={false}
    icon={[Function]}
  >
    <NavLink__Title>
      Page 1
    </NavLink__Title>
  </IconText>
</NavLink__Link>
`;

exports[`NavLink should show a badge when it has errors 1`] = `
.c0 {
  border-radius: 0.7em;
  background-color: #D0021B;
  color: white;
  padding: 0.2em 0.5em;
  font-weight: normal;
  z-index: 2;
  margin-left: auto;
  line-height: 1;
  font-size: 0.9rem;
  pointer-events: none;
}

<NavLink__Badge>
  <span
    className="c0"
  >
    2
  </span>
</NavLink__Badge>
`;
